File: Hebrew BOM/_build_scripts/scan_glosses.py
import re
import os
import sys

sys.stdout.reconfigure(encoding='utf-8')

filepath = os.path.join(os.environ.get('USERPROFILE', ''), 'Desktop', 'Hebrew BOM', 'BOM.html')
with open(filepath, 'r', encoding='utf-8') as f:
    lines = f.readlines()

# Comprehensive set of known proper names / legitimate single-cap words
legitimate = {
    # BOM proper names
    'Nephi','Alma','Moroni','Ammonihah','Zarahemla','Zeezrom','Amulek','Gideon',
    'Nehor','Mosiah','Helaman','Limhi','Abinadi','Noah','Laban','Lehi','Sariah',
    'Sam','Laman','Lemuel','Jerusalem','Israel','Moses','Aaron','Zoram','Ishmael',
    'Christ','Messiah','God','Lord','Zion','Joseph','Egypt','Jacob','Esau','David',
    'Solomon','Isaiah','Judah','Benjamin','Zedekiah','Isaac','Abraham','Bountiful',
    'Irreantum','Nahom','Liahona','Shazer','Coriantumr','Moronihah','Shiz','Ether',
    'Morianton','Pahoran','Teancum','Moriantum','Desolation','Cumorah','Ramah',
    'Ablom','Jared','Gilgal','Seth','Adam','Eve','Riplakish','Shiblon','Corianton',
    'Zeniff','Giddonah','Sidon','Melek','Jershon','Antionah','Zoramite','Rameumptom',
    'Onidah','Antionum','Amnigaddah','Amalickiah','Mormon','Amen','Selah','Joshua',
    'Gilead','Assyria','Babylon','Lebanon','Jordan','Samaria','Syria','Ephraim',
    'Manasseh','Rezin','Ahaz','Immanuel','Lucifer','Galilee','Midian','Nod','Bashan',
    'Jesus','Yeshua','Smith','America','Manchester','Ontario','New','York','Urim',
    'Thummim','Pharaoh','Hosanna','Nazareth','Counselor','Creator','Jeremiah',
    'Abarah','Judean','Enos','Jarom','Omni','Chemish','Amaron','Abinadom',
    'Helam','Shilom','Shemlon','Amulon','Sinim','Tarshish','Ophir','Chaldeans',
    'Philistia','Sheol','Uzziah','Aram','Damascus','Uriah','Jeberechiah',
    'Zebulun','Naphtali','Calno','Hamath','Oreb','Geba','Saul','Anathoth',
    'Madmenah','Gebim','Jesse','Coriantor','Moron','Sherem','Mary','Sinai',
    'Zenock','Neum','Zenos','Ammonites','Lamanite','Nephite','Zoramite',
    'Gidgiddoni','Lachoneus','Zemnarihah','Giddianhi','Kishkumen','Gadianton',
    'Timothy','Jonas','Mathoni','Mathonihah','Kumen','Kumenonhi',
    'Amaleki','Muloki','Ammah','Abish','Lamoni','Middoni','Antiomno',
    'Anti','Antipas','Zerahemnah','Amalickiahites',
    'Teomner','Antipus','Mulek','Judea','Cumeni','Antiparah','Manti',
    'Nephihah','Gid','Amalekites',
    'Hagoth','Tubaloth','Cezoram','Samuel','Lib','Shule','Kib','Orihah',
    'Corihor','Shared','Akish','Heth','Shez','Kim','Levi','Corom',
    'Omer','Esrom','Nimrod','Pagag','Com','Shiblom',
    'Coriantum','Ethem','Ogath','Ripliancum','Shurr','Cohor',
    'Jaredite','Zenephi','Neas','Sheum','Rahab','Ammon',
    # Legitimate English words used as glosses
    'He','Him','Himself','Eternal','Almighty','John','Reeds',
    'Holiness','Hosts','Bible','Jew',
    'Lamanites','Nephites','Jacobites','Josephites','Zoramites','Lemuelites',
    'Gentiles','Gentile','Hebrew','Lamb','Israelites','Me',
    'She','Her','His','My','Our','Your','Their','Its',
    'One','Two','Three','Jaredites','Amalekite',
    'Rabbanah','Sebus','Aminadab','Limnah','Onti','Seon','Shum',
    'Ezias','Isabel','Hermounts',
    'Zedekiah','Mulek','Sidon',
    # Other common English words that might appear capitalized
    'Sabbath','Passover','Tabernacle','Temple',
}

suspicious = []

for i in range(3134, min(47100, len(lines))):
    line = lines[i]
    pairs = re.findall(r'\["([^"]*?)","([^"]*?)"\]', line)
    for hebrew, gloss in pairs:
        if not gloss or gloss == '':
            continue

        flagged = False
        reason = ''

        # Pattern 1: Contains "???"
        if '???' in gloss:
            flagged = True
            reason = 'Contains ???'

        # Pattern 2: Single capitalized word that's NOT in legitimate names list
        # and looks like a Hebrew transliteration
        elif re.match(r'^[A-Z][a-z]+$', gloss) and gloss not in legitimate:
            lower = gloss.lower()
            # Check if it resembles a Hebrew transliteration vs. a real English word
            # Hebrew transliterations: unusual consonant combos, no standard English word form
            # Common English words we should NOT flag
            english_words = {
                'eternal','almighty','holiness','hosts','creator','counselor',
                'reeds','bible','jew','gentile','gentiles','lamb','hebrew',
                'me','he','him','she','her','his','my','our','your','their',
                'its','one','two','three','four','five','six','seven','eight','nine','ten',
                'north','south','east','west','king','queen','priest','prophet',
                'angel','spirit','heaven','hell','earth','water','fire','wind',
                'sword','shield','bow','arrow','stone','iron','brass','gold','silver',
                'copper','steel','wood','fruit','seed','tree','vine','olive',
                'wheat','barley','corn','wine','oil','honey','milk','bread',
                'city','land','sea','river','mountain','valley','wilderness',
                'desert','forest','field','garden','tower','wall','gate','door',
                'house','temple','altar','throne','judgment','covenant','law',
                'commandment','testimony','record','scripture','book','letter',
                'word','voice','name','people','nation','tribe','family',
                'father','mother','son','daughter','brother','sister','wife','husband',
                'child','infant','youth','man','woman','servant','master','lord',
                'ruler','captain','general','commander','chief','elder','judge',
                'war','peace','battle','army','host','camp','prison','death','life',
                'blood','flesh','bone','soul','heart','mind','eye','hand','foot',
                'head','mouth','ear','face','skin','hair',
                'righteous','wicked','holy','pure','clean','unclean',
                'true','false','good','evil','great','small','mighty','strong','weak',
                'rich','poor','wise','foolish','proud','humble','faithful','merciful',
                'just','unjust','free','captive',
                'love','hate','joy','sorrow','anger','fear','hope','faith',
                'grace','mercy','truth','light','darkness','glory','power',
                'salvation','redemption','atonement','resurrection','baptism',
                'repentance','forgiveness','blessing','curse','oath','promise',
                'sign','wonder','miracle','vision','dream','revelation',
                'offering','sacrifice','tithe','fast','prayer','worship',
                'morning','evening','night','day','month','year','time','season',
                'forever','always','never','now','then','here','there',
            }
            if lower not in english_words:
                flagged = True
                reason = 'Single capitalized word - possible transliteration'

        if flagged:
            suspicious.append((i+1, hebrew, gloss, reason))

for linenum, heb, gloss, reason in sorted(suspicious):
    print(f'Line {linenum}: ["{heb}","{gloss}"] - {reason}')

print(f'\nTotal suspicious: {len(suspicious)}')

File: Hebrew BOM/_build_scripts/scan_glosses.js
// Scan BOM.html for incorrectly glossed words (transliterations instead of English)
const fs = require('fs');
const bom = fs.readFileSync('BOM.html', 'utf8');

const re = /\["([^"]+)","([^"]*)"\]/g;
let m;
const suspicious = [];
let total = 0;

while ((m = re.exec(bom)) !== null) {
  const heb = m[1];
  const gloss = m[2];
  if (!gloss || gloss === '' || gloss === '׃') continue;
  total++;

  // Skip sof pasuk marker
  if (heb === '׃') continue;

  // Check if gloss contains Hebrew characters (definitely wrong)
  if (/[\u0590-\u05FF]/.test(gloss)) {
    suspicious.push({ hebrew: heb, gloss, reason: 'contains-hebrew-chars' });
    continue;
  }

  // Skip known good English patterns
  // Real glosses use lowercase-hyphenated like "and-it-was", "the-land", etc.
  // Or proper names like "Nephi", "Moroni"
  // Or prefix patterns like "[ACC]-the-word"

  // Transliterations are typically: capitalized, no hyphens, contain consonant clusters
  // like "Ohthlko", "Bhtbonno", "Nchlko", "Tbshlo", etc.
  // They come from the transliterate() function which capitalizes first letter

  // Check for likely transliteration patterns:
  // 1. Starts with uppercase, rest is lowercase, no hyphens, and has unusual consonant clusters
  const isSimpleTranslit = /^[A-Z][a-z]+$/.test(gloss) &&
    !/^(Nephi|Moroni|Mormon|Mosiah|Alma|Helaman|Ether|Lehi|Jacob|Enos|Jarom|Omni|Ammon|Aaron|Coriantumr|Limhi|Zeniff|Noah|Abinadi|Gideon|Amulek|Zeezrom|Korihor|Pahoran|Teancum|Lachoneus|Giddianhi|Gidgiddoni|Kishkumen|Gadianton|Amalickiah|Lamoni|Abish|Jared|Nimrod|Omer|Heth|Shule|Riplakish|Morianton|Lib|Shiz|Gilgal|Bountiful|Zarahemla|Melek|Ammonihah|Gideon|Jershon|Manti|Mulek|Cumorah|Desolation|Abundance|Jerusalem|Eden|Sidon|Israel|Ishmael|Laman|Lemuel|Sam|Sariah|Zoram|Hagoth|Shiblon|Corianton|Timothy|Jonas|Mathoni|Mathonihah|Kumen|Kumenonhi|Jeremiah|Shemnon|Zedekiah|Nephi|Isaiah|Christ|Jesus|Mary|Joseph|Moses|Adam|Eve|Abel|Cain|Enoch|David|Solomon|Elijah|Malachi|Samuel|Abigail|Sarah|Abraham|Seth|God|Lord|Zion|Egypt|Babylon)$/.test(gloss);

  if (isSimpleTranslit) {
    // Further check: does it have consonant clusters unlikely in English?
    // English proper names don't usually have "tsh", "khl", "shm", "dch", etc.
    const lower = gloss.toLowerCase();
    const hasWeirdCluster = /[bcdfghjklmnpqrstvwxyz]{4,}/.test(lower) ||
      /^[bcdfghjklmnpqrstvwxyz]{2}[aeiou]/.test(lower) ||
      /tsh|khl|shm|dch|tsb|bsh|nbl|gvr|mkh|shp|shk|shl|tbl|tkh|pkl|sht/.test(lower);
    if (hasWeirdCluster || lower.length <= 4) {
      suspicious.push({ hebrew: heb, gloss, reason: 'likely-transliteration' });
      continue;
    }
  }

  // 2. Mixed case in middle (like "Oitbonn", "Bhtbonno")
  if (/^[A-Z][a-z]*[A-Z]/.test(gloss) && !gloss.includes('-')) {
    suspicious.push({ hebrew: heb, gloss, reason: 'mixed-case-transliteration' });
    continue;
  }

  // 3. Hyphenated but contains a transliteration segment
  // e.g., "and-Nchlko" or "[ACC]-Tbshlo"
  const parts = gloss.split(/[-]/);
  for (const part of parts) {
    if (/^[A-Z][a-z]+$/.test(part) && part.length >= 3) {
      const lower = part.toLowerCase();
      const hasWeirdCluster = /[bcdfghjklmnpqrstvwxyz]{4,}/.test(lower) ||
        /tsh|khl|shm|dch|tsb|bsh|nbl|gvr|mkh|shp|shk|tbl|tkh|pkl/.test(lower);
      if (hasWeirdCluster) {
        suspicious.push({ hebrew: heb, gloss, reason: 'contains-transliterated-part' });
        break;
      }
    }
  }
}

// Deduplicate by gloss
const glossMap = {};
for (const s of suspicious) {
  const key = s.gloss;
  if (!glossMap[key]) glossMap[key] = { hebrew: s.hebrew, gloss: s.gloss, reason: s.reason, count: 0 };
  glossMap[key].count++;
}

const sorted = Object.values(glossMap).sort((a, b) => b.count - a.count);
console.log('Total glosses scanned:', total);
console.log('Suspicious/incorrect glosses found:', sorted.length, 'unique,', suspicious.length, 'total occurrences');
console.log('\n--- All suspicious glosses ---');
sorted.forEach(s => console.log(`  ${s.gloss} (${s.hebrew}) x${s.count} [${s.reason}]`));

fs.writeFileSync('suspicious_glosses.json', JSON.stringify(sorted, null, 2));
